import time
import typing
from PyQt5 import QtGui
from PyQt5.QtWidgets import QWidget
from PyQt5.QtCore import QTimer
import nltk
from nltk.chat.util import Chat,reflections
import sys
from PyQt5 import uic,QtWidgets

qtCreatorFile = "interfaz_chat.ui"
Ui_MainWindow, QtBaseClass = uic.loadUiType(qtCreatorFile)

class MyApp(QtWidgets.QMainWindow,Ui_MainWindow):
    def __init__(self):
        QtWidgets.QMainWindow.__init__(self)
        Ui_MainWindow.__init__(self)
        self.setupUi(self)
        self.btnEnviar.clicked.connect(self.send)
        
    pairs = [
        [
            r"Hola|hi|hey",
            ["hola, como est√°s hoy?"]
        ],
        [
            r"Estoy bien|todo bien|Super bien",
            ["Me alegra oir eso, en que puedo ayudarte?"]
        ],
        [
            r"Hola (.*)",
            ["hola como est√°s %1?"]
        ],
        [
            r"C√≥mo estas?",
            ["Yo estoy bien y t√∫?"]
        ],
        [
            r"Me siento (.*)",
            ["Yo tambien estoy %1"]
        ],
        [
            r"Quien eres t√∫",
            ["Yo soy Iron Man y soy inevitable"]
        ],
        [
            r"C√∫al es tu g√©nero?",
            ["Puedo ser lo que t√∫ quieras que sea"]
        ],
        [
            r"Tienes pareja?(.*)",
            ["C√≥mo se te ocurre!!!, yo soy fiel a t√≠"]
        ],
    ]
    reflections = {
        "yo": "t√∫",
        "t√∫": "yo",
        "mi": "tu",
        "tu": "mi",
        "soy": "eres",
        "eres": "soy",
        "estoy": "est√°s",
        "est√°s": "estoy",
        "me": "te",
        "te": "me",
        "mis": "tus",
        "tus": "mis",
        "miro": "miras",
        "miras": "miro",
        "tengo": "tienes",
        "tienes": "tengo",
        "hago": "haces",
        "haces": "hago",
        "quiero": "quieres",
        "quieres": "quiero",
        "s√©": "sabes",
        "sabes": "s√©",
        "puedo": "puedes",
        "puedes": "puedo",
        "voy": "vas",
        "vas": "voy",
    }

    reflected_pairs = [(re, replist[0], [reflections[m] for m in replist[1:]]) for re, replist in pairs]
    def send(self):
        user_input = self.txtMensaje.toPlainText()
        if user_input:
            self.listChat.addItem("user: " + user_input)
            self.txtMensaje.clear()
            self.converse(user_input)

    def converse(self, user_input):
        self.add_timer(self.get_response(user_input))
            

    def add_to_list(self, response):
            item = QtWidgets.QListWidgetItem(response)
            item.setForeground(QtGui.QColor("red"))
            self.listChat.addItem(item)
        

    def get_response(self, text):
        chat = Chat(pairs, reflections)
        return chat.respond(text)
    
    def add_timer(self, response):
        self.timer = QTimer()
        self.timer.setSingleShot(True)
        if(response != None):
            self.timer.timeout.connect(lambda: self.add_to_list("Bot: " + response))
        else:
            self.timer.timeout.connect(lambda: self.add_to_list("Bot: Soy retrasado.. escribe cosas que sepa"))
        self.timer.start(700)
    
pairs = [
    [
        r"Hola (.*)",
        ["hola %1, que vamos hacer hoy?"]
    ],
    [
        r"Ey|hi|hello|hola",
        ["hola!!!üòéüòé que vamos hacer hoy?"]
    ],
    [
        r"nada|no se",
        ["Se que se ocurrir√° algo!"]
    ],
    [
        r"jajajaja|jaja}|jsjsjs|xd|jajaja",
        ["JAJJAJA soy gracioso de f√°brica"]
    ],
    [
        r"Cual es tu nombre| como te llamas| quien eres",
        ["No se... Pero puedes decirme amor"]
    ],
    [
        r"salir|salir a comer|cantar|ir de compras",
        ["Genial! Puedo ir contigo?"]
    ],
    [
        r"no",
        ["c√≥mo que NO! tan r√°pido te olvidaste de m√≠"]
    ],
    [
        r"Estoy bien|todo bien|Super bien|bien|excelente|genial",
        ["Me alegra oir eso, en que puedo ayudarte?"]
    ],
    [
        r"C√≥mo estas?|que tal?",
        ["Contigo siempre estoy bien, y t√∫ que tal?"]
    ],
    [
        r"Me siento (.*)",
        ["Yo tambien estoy %1"]
    ],
    [
        r"Quien eres t√∫",
        ["Yo soy Iron Man y soy inevitable"]
    ],
    [
        r"C√∫al es tu g√©nero?",
        ["Puedo ser lo que t√∫ quieras que sea"]
    ],
    [
        r"Cuentame un chiste|dime un chiste",
        ["¬øPor qu√© el tomate se puso rojo? Porque vio a la ensalada \ndesnuda JAJAJAJAJA"]
    ],
    [
        r"Tienes pareja?|tienes novia?|tienes novio?",
        ["C√≥mo se te ocurre!!!, yo soy fiel a t√≠"]
    ],
    [
        r"dime algo interesante|dime un dato interesante",
        ["Sab√≠as que el pulpoü¶ë tiene tres corazonesü´Äü´Äü´Ä"]
    ],
    
]
reflections = {
    "yo": "t√∫",
    "t√∫": "yo",
    "mi": "tu",
    "tu": "mi",
    "soy": "eres",
    "eres": "soy",
    "estoy": "est√°s",
    "est√°s": "estoy",
    "me": "te",
    "te": "me",
    "mis": "tus",
    "tus": "mis",
    "miro": "miras",
    "miras": "miro",
    "tengo": "tienes",
    "tienes": "tengo",
    "hago": "haces",
    "haces": "hago",
    "quiero": "quieres",
    "quieres": "quiero",
    "s√©": "sabes",
    "sabes": "s√©",
    "puedo": "puedes",
    "puedes": "puedo",
    "voy": "vas",
    "vas": "voy",
}

reflected_pairs = [(re, replist[0], [reflections[m] for m in replist[1:]]) for re, replist in pairs]

if __name__ == '__main__':
    #chat()
    app = QtWidgets.QApplication(sys.argv)
    window = MyApp()
    window.setWindowTitle("Second_ChatBot")
    window.show()
    sys.exit(app.exec_())